### DISCUSSION TITLE
Quick Start ?'s

### DISCUSSION THREAD

**Discussion by hopsj** [usnistgov/NEMO#63] [Q&A] [answered]
Hello, I'm attempting to run through the install, figured I'd try the [quick start ](https://github.com/usnistgov/NEMO/wiki/NEMO-with-docker-Quick-Start) first.  Worked a charm at home on my mac to get the example up and running, but running on a Linux server at work (Red Hat Enterprise Linux release 8.3 (Ootpa)), the startup `./start.sh` fails, ending with an error message `ModuleNotFoundError: No module named 'settings'` (full output below).  I'm new to docker, and not sure where to start on troubleshooting, any advice?

Full output:

```
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Completed short name "nanofab/nemo" with unqualified-search registries (origin: /etc/containers/registries.conf)
Trying to pull registry.access.redhat.com/nanofab/nemo:1.12.0...
  name unknown: Repo not found
Trying to pull registry.redhat.io/nanofab/nemo:1.12.0...
  unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication
Trying to pull docker.io/nanofab/nemo:1.12.0...
Getting image source signatures
Copying blob 6fba9447437b done
Copying blob 400bad26d6c0 done
Copying blob 22dbe790f715 done
Copying blob c2b4d327b352 done
Copying blob 270e1baa5299 done
Copying blob 0250231711a0 done
Copying blob 1224d2c3c8d9 done
Copying blob 1d360410c080 done
Copying blob 5d7d0629cdba done
Copying blob fc4f27ca3564 done
Copying blob 7328668264ea done
Copying blob a552e5eeb4a3 done
Copying blob 837b0a6116c4 done
Copying blob 75889c533605 done
Copying blob 23f017732bcf done
Copying blob b94275762297 done
Copying config 226f121fe1 done
Writing manifest to image destination
Storing signatures
226f121fe13881942ad178e646e036d1da5adcfccd2744c071239f99f905a46e
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Completed short name "nanofab/nginx" with unqualified-search registries (origin: /etc/containers/registries.conf)
Trying to pull registry.access.redhat.com/nanofab/nginx:2.1.0...
  name unknown: Repo not found
Trying to pull registry.redhat.io/nanofab/nginx:2.1.0...
  unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication
Trying to pull docker.io/nanofab/nginx:2.1.0...
Getting image source signatures
Copying blob 1e437faeaaf7 done
Copying blob 7bd5fd7982df done
Copying blob a02a4930cb5d done
Copying blob 34d3cf8192a3 done
Copying blob 9c17ba3e9d22 done
Copying blob e4d14390a45d done
Copying blob 281592774a5d done
Copying blob e05db17b1635 done
Copying blob d088ffa2c21a done
Copying blob 38823a41998f done
Copying blob 7c33c117557d done
Copying blob cdcf67353299 done
Copying blob 59b5eaa7baf9 done
Copying blob 214e07818daa done
Copying config 3871b84010 done
Writing manifest to image destination
Storing signatures
3871b84010f5e076bc338718f5f63a127ae78e1278ff4d0ed8cd55b772c67328
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Completed short name "nanofab/control_center" with unqualified-search registries (origin: /etc/containers/registries.conf)
Trying to pull registry.access.redhat.com/nanofab/control_center:2.3.0...
  name unknown: Repo not found
Trying to pull registry.redhat.io/nanofab/control_center:2.3.0...
  unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication
Trying to pull docker.io/nanofab/control_center:2.3.0...
Getting image source signatures
Copying blob a02a4930cb5d skipped: already exists
Copying blob c475cd8cac66 done
Copying blob dc1dbca7249b done
Copying blob d392c79f6210 done
Copying blob bc8db6a2cf42 done
Copying blob 3a64e5a1a463 done
Copying blob c30b07bbedd4 done
Copying blob eb31925cba6c done
Copying blob 32b0fabed8e5 done
Copying blob 957cd90c4a87 done
Copying blob dfeff1c1d43a done
Copying blob 967ec5d8884a done
Copying blob 2f6b4d811f79 done
Copying blob 2ec0e6fe18be done
Copying config 72124c3202 done
Writing manifest to image destination
Storing signatures
72124c320231fb7985158a130c88256da91c7d63649a8cc07197b5a9bd28ac52
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
Traceback (most recent call last):
  File "/usr/local/bin/django-admin", line 10, in <module>
    sys.exit(execute_from_command_line())
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 308, in execute
    settings.INSTALLED_APPS
  File "/usr/local/lib/python3.6/site-packages/django/conf/__init__.py", line 56, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python3.6/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/local/lib/python3.6/site-packages/django/conf/__init__.py", line 110, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'settings'
```

**Comment by rptmat57** [comment]
Hello,
This seems to suggest the `settings.py` is not present where it's supposed to be <pwd>/nemo/settings.py
Can you check that it is there?
The failing command is `docker run --interactive --tty --volume $(pwd)/nemo:/nemo nanofab/nemo:1.12.0 django-admin migrate`, so you can try running that one manually and see if it works.

**Reply by hopsj** [reply]
Thanks for the reply.  Yes settings.py is present in the ./nemo subdirectory from the extracted zip file.  Running the command manually produces the same result.  

Can you help me understand those error messages some?  It seems the python call stack is going through `/usr/local/lib/python3.6/site-packages/django/...` .  Are those paths referenced from within the docker image, or do they refer to the filesystem of the host machine itself?  (The host machine does have django installed, but I'm not very familiar with either docker or django)

**Reply by rptmat57** [reply]
those path are referenced from within the docker image.
try replacing $(pwd) by the absolute path.
for example: `docker run --interactive --tty --volume /opt/nemo:/nemo nanofab/nemo:1.12.0 django-admin migrate`

**Reply by hopsj** [reply]
still no dice.  I've got an appt w/ one of our IT experts later, will copy back here if I find anything.  One thing I noticed in the output is: `Emulate Docker CLI using podman`.  I guess that means I'm not actually using docker, but rather podman?  I did some googling and [see that apparently docker is not supported on RHEL 8](https://access.redhat.com/solutions/3696691).  Perhaps an artifact of using podman vs docker?  I might be able to get a system up using a different flavor of linux.

**Reply by rptmat57** [reply]
I have not tried on RHEL 8. However we use RHEL 7 and I have not had any issues with it.
You could try:
`docker run -it -v $(pwd)/nemo:/nemo nanofab/nemo:1.12.0 ls -al /nemo`
to see if the folder is indeed mounted into the container and that you can see settings.py

FYI The script and NEMO versions are way outdated (we are at 3.7.0 currently) but it shouldn't really affect things (I need to update it at some point)

**Reply by hopsj** [reply]
Bit of an update here.  If I use a oneliner command:
`docker run --detach --name name_splash_pad --publish 8000:8000 nanofab/nemo_splash_pad`

Seems to work OK.

My going theory is above docker is not mounting the volume properly?  I suppose that's a docker issue then? I tried running just 'ls' command, but not sure if I'm using the docker run command correctly:

```
docker run --volume /home/<username>/nemo:/nemo nanofab/nemo ls
ls: cannot open directory '.': Permission denied
```

**Reply by rptmat57** [reply]
oh it's a permission issue. by default NEMO's docker container is run as root.
so you'll need to run it as root for it to work.

**Reply by hopsj** [reply, marked as answer]
Thanks again for the reply, tried running via sudo, but no dice. 

However, now after two days of googling, I seem to have found a solution!  See https://devops.stackexchange.com/questions/11267/permission-denied-within-mounted-volume-inside-docker-podman-container

Nutshell is adding `:z` at the end of the the volume.  Not sure yet what exactly this means, but alas the ls command is at least working:
```
docker run --volume /home/<username>/nemo:/nemo:z nanofab/nemo ls
logs
media
secrets
settings.py
static
```

**Reply by rptmat57** [reply]
Ok I am glad this is working now. 
It might be helpful for others trying to run NEMO on RHEL 8 / podman!
